packages/worker/src/worker.ts

```diff
@@ -854,85 +856,79 @@ export class Worker {
+                      );
+                    }
+                    const info = await extractActivityInfo(
+                      task,
+                      this.options.loadedDataConverter,
+                      this.options.namespace,
+                      this.options.taskQueue
+                    );
+
+                    const { activityType } = info;
+                    const fn = this.options.activities.get(activityType);
+                    if (typeof fn !== 'function') {
+                      throw ApplicationFailure.create({
+                        type: 'NotFoundError',
+                        message: `Activity function ${activityType} is not registered on this Worker, available activities: ${JSON.stringify(
+                          [...this.options.activities.keys()]
+                        )}`,
+                        nonRetryable: false,
+                      });
+                    }
+                    let args: unknown[];
+                    try {
+                      args = await decodeArrayFromPayloads(this.options.loadedDataConverter, task.start?.input);
+                    } catch (err) {
+                      throw ApplicationFailure.fromError(err, {
+                        message: `Failed to parse activity args for activity ${activityType}: ${errorMessage(err)}`,
+                        nonRetryable: false,
+                      });
+                    }
+                    const headers = task.start?.headerFields ?? {};
+                    const input = {
+                      args,
+                      headers,
                     };
+
+                    this.logger.trace('Starting activity', activityLogAttributes(info));
+
+                    activity = new Activity(
+                      info,
+                      fn,
+                      this.options.loadedDataConverter,
+                      (details) =>
+                        this.activityHeartbeatSubject.next({
+                          type: 'heartbeat',
+                          info,
+                          taskToken,
+                          base64TaskToken,
+                          details,
+                          onError() {
+                            activity?.cancel('HEARTBEAT_DETAILS_CONVERSION_FAILED'); // activity must be defined
+                          },
+                        }),
+                      this.logger,
+                      this.options.interceptors.activity
+                    );
+                    output = { type: 'run', activity, input };
                     break;
-                  }
-                  let args: unknown[];
-                  try {
-                    args = await decodeArrayFromPayloads(this.options.loadedDataConverter, task.start?.input);
-                  } catch (err) {
+                  } catch (e) {
+                    const error = ensureApplicationFailure(e);
                     output = {
```

Comment:
What happens if it is not an application failure? Can you add a comment?

Reply:
`ensureApplicationFailure` is an already existing internal helper function. 

```
/**
 * If `error` is already an `ApplicationFailure`, returns `error`.
 *
 * Otherwise, converts `error` into an `ApplicationFailure` with:
 *
 * - `message`: `error.message` or `String(error)`
 * - `type`: `error.constructor.name` or `error.name`
 * - `stack`: `error.stack` or `''`
 */
export function ensureApplicationFailure(error: unknown): ApplicationFailure {
```

I don't there's more to say than this. WDYT?

Reply:
I thought that was an assertion, never mind...